zebrad/src/components/inbound/tests/real_peer_set.rs

```diff
@@ -0,0 +1,498 @@
+//! Inbound service tests with a real peer set.
+
+use std::{collections::HashSet, iter, net::SocketAddr, sync::Arc};
+
+use futures::FutureExt;
+use tokio::{sync::oneshot, task::JoinHandle};
+use tower::{
+    buffer::Buffer,
+    builder::ServiceBuilder,
+    util::{BoxCloneService, BoxService},
+    ServiceExt,
+};
+
+use zebra_chain::{
+    block::{self, Block},
+    parameters::Network,
+    transaction::{AuthDigest, Hash as TxHash, UnminedTxId, WtxId},
+};
+use zebra_consensus::{chain::VerifyChainError, error::TransactionError, transaction};
+use zebra_network::{
+    connect_isolated_tcp_direct, Config as NetworkConfig, Request, Response, ResponseStatus,
+    SharedPeerError,
+};
+use zebra_state::Config as StateConfig;
+use zebra_test::mock_service::{MockService, PanicAssertion};
+
+use crate::{
+    components::{
+        inbound::{Inbound, InboundSetupData},
+        mempool::{self, gossip_mempool_transaction_id, Mempool},
+        sync::{self, BlockGossipError, SyncStatus},
+    },
+    BoxError,
+};
+
+use ResponseStatus::*;
+
+/// Check that a network stack with an empty address book only contains the local listener port,
+/// by querying the inbound service via a local TCP connection.
+///
+/// Uses a real Zebra network stack with a local listener address,
+/// and an isolated Zebra inbound TCP connection.
+#[tokio::test]
+async fn inbound_peers_empty_address_book() -> Result<(), crate::BoxError> {
```

Comment:
Minor: I think the function doesn't need to return a `Result`:

```suggestion
async fn inbound_peers_empty_address_book() {
```

Reply:
I don't think this matters that much in tests.